zone/client_packet.cpp

```diff
@@ -14765,3 +14773,177 @@ void Client::Handle_OP_ResetAA(const EQApplicationPacket *app)
 	}
 	return;
 }
+
+void Client::CheatDetected(CheatTypes CheatType, float x, float y, float z)
+{
+	switch (CheatType)
+	{
+	case MQWarp:
+		if (RuleB(Zone, EnableMQWarpDetector) && ((this->Admin() < RuleI(Zone, MQWarpExemptStatus) || (RuleI(Zone, MQWarpExemptStatus)) == -1)))
+		{
+			Message(13, "Large warp detected.");
+			char hString[250];
+			sprintf(hString, "/MQWarp with location %.2f, %.2f, %.2f", GetX(), GetY(), GetZ());
+			database.SetMQDetectionFlag(this->account_name, this->name, hString, zone->GetShortName());
+		}
+		break;
+	default:
+		char* hString = nullptr;
+		MakeAnyLenString(&hString, "Unhandled HackerDetection flag with location %.2f, %.2f, %.2f.", GetX(), GetY(), GetZ());
+		database.SetMQDetectionFlag(this->account_name, this->name, hString, zone->GetShortName());
+		safe_delete_array(hString);
+		break;
+	}
+}
+
+bool Client::IsMQExemptedArea(uint32 zoneID, float x, float y, float z)
```

Comment:
Instead of "exempt areas" I would prefer to have coverage over "exempt behaviors"

For example if we are performing a door click and warping, or performing a normal action and warping, we need to be able to trigger the timer to let the server consider that you are performing normal behavior. Manually specifying zones and areas is far too much manual work and maintainability that is prone to producing false positives